Surface Depth Facet

MaxBorsch opened the pull request:
1. Added an optional surface depth facet that when added limits the default surface rasterizer from generating below the depth.

2. Made spawning point selection return a value even if no "good" spawn could be found, because it's better to at least spawn in the ground than crash the game.

Part of GCI task creating an Underworld.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Cervator commented:
I can confirm this lets Perlin + Underworld create a secondary world layer for the underworld :-)

I'm not sure about the longer search for a spawn position. I think I'd prefer timing out and rather than having that stall/crash the game return gracefully to the main menu with a sensible error message.

Forcing a spawn (possibly after an excessive amount of time searching a spiral of max size 250?) in an invalid position could be very confusing for a new player. Better to have a clean error message :-)

And not that it is a big deal but you've got a 2014 license header in there :D

Haven't tested a lot of other worlds yet (sleepy). Might be good with some bits of review from elsewhere so throwing a few reviewers on here.

skaldarnar commented on engine/src/main/java/org/terasology/logic/players/PlayerFactory.java:
```diff
@@ -134,15 +134,15 @@ private Vector3f findSpawnPos(Vector3f targetPos, float entityHeight) {
         int targetBlockZ = TeraMath.floorToInt(targetPos.z);
         Vector2i center = new Vector2i(targetBlockX, targetBlockZ);
-        for (BaseVector2i pos : SpiralIterable.clockwise(center).maxRadius(3).scale(2).build()) {
+        for (BaseVector2i pos : SpiralIterable.clockwise(center).maxRadius(250).scale(2).build()) {
```
3 might be a bit low, but 250 seems too high. Maybe 16 or 32 is a better compromise? What is the default chunk size nowadays?

emanuele3d replied:
32x64x32 if I remember correctly.

skaldarnar commented on engine/src/main/java/org/terasology/logic/players/PlayerFactory.java:
```diff
@@ -134,15 +134,15 @@ private Vector3f findSpawnPos(Vector3f targetPos, float entityHeight) {
             }
         }
-        return null;
+        return targetPos;
```
Should we go for `Optional[Vector3i]̀  as return type?

See: https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

skaldarnar commented on engine/src/main/java/org/terasology/world/generation/facets/SurfaceDepthFacet.java:
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2014 MovingBlocks
```
2014 license header.

skaldarnar commented on engine/src/main/java/org/terasology/world/generation/facets/SurfaceDepthFacet.java:
```diff
@@ -0,0 +1,30 @@
+import org.terasology.world.generation.facets.base.BaseFieldFacet2D;
+
+/**
+ * Stores the surface depth limits
```
Please briefly explain what the "surface depth" actually is…

MaxBorsch commented:
@skaldarnar I fixed all the stuff you mentioned. Made the getSpawn() return an optional. Added a TODO for handling an empty optional, since that is a whole other task and we haven't really agreed on what should happen.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Cervator commented:
Retested, this time used railgun to drill a tunnel straight down .... woo! We might need to increase the cooldown a bit on that thing.

Added a comment to clarify what the new line does in SolidRasterizer. Also checked out Polyworld just in case (different player spawn location picking)

Will submit a follow-up for figuring out a player spawn timeout. Anybody else with comments please submit here or on the new issue :-)

VaibhavBajaj commented:
@MaxBorsch Maybe the ceiling of this world should be mantleRock so that no matter how crazy a guy is using a rail gun, he can't reach the bottom. Just a suggestion.

Note: MantleRock is indestructible.

MaxBorsch commented:
@VaibhavBajaj You're supposed to be able to dig to the bottom. Also the underworld then has no bottom.

VaibhavBajaj commented:
@MaxBorsch Then why the portals? Also, why would the mantle rock disappear from the bottom? I was suggesting adding it in your WorldRasterizer, not actually changing the position of the whole Underground to below the mantle rock.

MaxBorsch commented:
@VaibhavBajaj The portals are useless, just a demo for portals, or to help people discover the underworld. Ideally they'd spawn in some necromancer castle etc.
